sdktests/server_side_events_custom.go

```diff
@@ -0,0 +1,128 @@
+	"gopkg.in/launchdarkly/go-sdk-common.v2/ldvalue"
+)
+
+func doServerSideCustomEventTests(t *ldtest.T) {
+	// These do not include detailed tests of the encoding of user attributes in custom events,
+	// which are in server_side_events_users.go.
+
+	t.Run("data and metricValue parameters", doServerSideParameterizedCustomEventTests)
+
+	t.Run("index events are created for new users if not inlined", func(t *ldtest.T) {
+		eventsConfig := baseEventsConfig()
+
+		dataSource := NewSDKDataSource(t, mockld.EmptyServerSDKData())
+		events := NewSDKEventSink(t)
+		client := NewSDKClient(t, WithConfig(servicedef.SDKConfigParams{Events: &eventsConfig}), dataSource, events)
+
+		user1 := lduser.NewUserBuilder("user1").Name("Ann").Build()
+		user2 := lduser.NewUserBuilder("user2").Name("Bob").Build()
+		user3 := lduser.NewUserBuilder("user3").Name("Cat").Build()
+
+		event1 := servicedef.CustomEventParams{EventKey: "event1", User: &user1} // generates index event for user1
+		event2 := servicedef.CustomEventParams{EventKey: "event2", User: &user1} // no new index, user1 already seen
+		event3 := servicedef.CustomEventParams{EventKey: "event3", User: &user2} // generates index event for user2
+		event4 := servicedef.CustomEventParams{EventKey: "event4", User: &user1} // no new index, user1 already seen
+		event5 := servicedef.CustomEventParams{EventKey: "event5", User: &user2} // no new index, user2 already seen
+		event6 := servicedef.CustomEventParams{EventKey: "event6", User: &user3} // generates index event for user3
+
+		for _, params := range []servicedef.CustomEventParams{event1, event2, event3, event4, event5, event6} {
+			client.SendCustomEvent(t, params)
+		}
+		client.FlushEvents(t)
+		payload := events.ExpectAnalyticsEvents(t, defaultEventTimeout)
+
+		expected := make([]m.Matcher, 0)
+		for _, params := range []servicedef.CustomEventParams{event1, event2, event3, event4, event5, event6} {
+			expected = append(expected, EventIsCustomEventForParams(params, eventsConfig))
+		}
+		for _, user := range []lduser.User{user1, user2, user3} {
+			expected = append(expected, EventIsIndexEvent(mockld.ExpectedEventUserFromUser(user, eventsConfig)))
+		}
+		m.AssertThat(t, payload, m.ItemsInAnyOrder(expected...))
+	})
+}
+
+func doServerSideParameterizedCustomEventTests(t *ldtest.T) {
+	eventsConfig := baseEventsConfig()
+	eventsConfig.InlineUsers = true // so we don't get index events in the output
+
+	dataSource := NewSDKDataSource(t, mockld.EmptyServerSDKData())
+	events := NewSDKEventSink(t)
+	client := NewSDKClient(t, WithConfig(servicedef.SDKConfigParams{Events: &eventsConfig}), dataSource, events)
+
+	user := lduser.NewUser("user-key")
+
+	// Generate many permutations of 1. data types that can be used for the data parameter, if any, and
+	// 2. metric value parameter, if any.
+	allParams := make([]servicedef.CustomEventParams, 0)
+	omitMetricValue := float64(-999999) // magic value that we'll change to null
```

Comment:
I wonder if https://pkg.go.dev/testing/quick could be utilized at all here? Perhaps it doesn't fit the use-case.

Reply:
Well... I see what you're getting at, but I'm not really trying to do black-box testing here in that sense. Maybe it's worth talking (outside of this PR) about whether we should do more in that regard, but in the current overall design of these tests I did make a firm decision not to go that route.

Pure black-box testing, when numeric or string parameters are involved, would imply running millions or billions of permutations for basic operations even when there's no _functional_ difference between the values. That is, in an operation where the value isn't used in any computation that could ever produce an error for value A as opposed to value B—for instance, if the value is simply being copied unchanged into the output—in pure black-box testing you're not allowed to know any such implementation details, so you still need to do just as many tests.

But if you're allowed to make at least very basic assumptions about what kinds of differences in behavior for different inputs might be plausible, and about what layer of the software you are trying to test, and what kinds of variances are or are not meaningful, then a more focused approach is possible. The assumptions being made in this case with custom events are:

* The `data` and `metricValue` parameters, if present, should cause those same values to be copied into the output event. They are not interpreted by SDKs in any other way.
* Event data is JSON data. The `data` parameter can be any JSON value. JSON is already a well-defined spec and if the overall implementation of JSON being used by the SDK was broken, that is a bigger problem that we'd expect to detect in tests that aren't specific to custom events. So in this test we aren't concerned with, for instance, a hypothetical error where the string "abc" gets encoded correctly but the string "abcd" does not.
* However, on most platforms, JSON data has to be converted to and from some built-in data types. So it is possible that an SDK could mistakenly declare `data` as a string, in which case it would work correctly for all string values but fail if `data` was a JSON number. Therefore it makes sense to test separate cases where `data` is each of the allowable JSON types.
* Similarly, the semantics of `metricValue` are that it is either a number (any number) or "no value". It's certainly possible that an SDK could handle one of those cases correctly but not the other, so we should test both.
* It is theoretically possible that an SDK could work for one numeric value of `metricValue` but not another, even though (in terms of how this method is defined) there's no reason for there to be any computations with it at all. But if our test is "the same number comes out that we put in", we can't blindly test it with every possible `float64` value, because JSON itself couldn't pass such a test— many `float64` numbers can't be represented accurately in a decimal format. That's a well-known limitation of any system that uses JSON numbers, and LaunchDarkly simply can't guarantee that kind of accuracy.

So, my compromise on that last point was to use just a couple of example numeric values which are meant to catch possible misuse of strong typing in the SDK logic. That is, even if the SDK doesn't do any computations with the value, if it mistakenly used an integer type or an unsigned type then it would mangle non-integer or negative values. Or if it assumed that zero meant "no value" (i.e. it didn't use a nullable type) then you wouldn't be able to send a zero value, so there's a test case for that.